config/versions/v4.go

```diff
@@ -0,0 +1,87 @@
+package versions
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/buger/jsonparser"
+)
+
+var (
+	errUpgradingAssetTypes    = errors.New("error upgrading assetTypes")
+	errUpgradingCurrencyPairs = errors.New("error upgrading currencyPairs.pairs")
+)
+
+// Version4 is an Exchange upgrade to move currencyPairs.assetTypes to currencyPairs.pairs.*.assetEnabled
+type Version4 struct {
+}
+
+func init() {
+	Manager.registerVersion(4, &Version4{})
+}
+
+// Exchanges returns all exchanges: "*"
+func (v *Version4) Exchanges() []string { return []string{"*"} }
+
+// UpgradeExchange sets AssetEnabled: true for all assets listed in assetTypes, and false for any with no field
+func (v *Version4) UpgradeExchange(_ context.Context, e []byte) ([]byte, error) {
+	toEnable := map[string]bool{}
+
+	assetTypesFn := func(asset []byte, valueType jsonparser.ValueType, _ int, _ error) {
+		if valueType == jsonparser.String {
+			toEnable[string(asset)] = true
+		}
+	}
+	_, err := jsonparser.ArrayEach(e, assetTypesFn, "currencyPairs", "assetTypes")
+	if err != nil && !errors.Is(err, jsonparser.KeyPathNotFoundError) {
+		return e, fmt.Errorf("%w: %w", errUpgradingAssetTypes, err)
+	}
+
+	assetEnabledFn := func(assetBytes []byte, v []byte, _ jsonparser.ValueType, _ int) (err error) {
+		asset := string(assetBytes)
+		if toEnable[asset] {
+			e, err = jsonparser.Set(e, []byte(`true`), "currencyPairs", "pairs", asset, "assetEnabled")
+		} else {
+			var vT jsonparser.ValueType
+			_, vT, _, err = jsonparser.Get(v, "assetEnabled")
+			switch {
+			case vT == jsonparser.Null, errors.Is(err, jsonparser.KeyPathNotFoundError):
+				e, err = jsonparser.Set(e, []byte(`false`), "currencyPairs", "pairs", asset, "assetEnabled")
+			case err == nil && vT != jsonparser.Boolean:
+				err = fmt.Errorf("assetEnabled: %w (`%s`)", jsonparser.UnknownValueTypeError, vT)
+			}
+		}
+		if err != nil {
+			err = fmt.Errorf("%w for asset `%s`", err, asset)
+		}
+		return err
+	}
+	if err = jsonparser.ObjectEach(bytes.Clone(e), assetEnabledFn, "currencyPairs", "pairs"); err != nil {
+		return e, fmt.Errorf("%w: %w", errUpgradingCurrencyPairs, err)
+	}
+	e = jsonparser.Delete(e, "currencyPairs", "assetTypes")
+	return e, err
+}
+
+// DowngradeExchange moves AssetEnabled assets into AssetType field
+func (v *Version4) DowngradeExchange(_ context.Context, e []byte) ([]byte, error) {
+	assetTypes := []string{}
+
+	assetEnabledFn := func(asset []byte, v []byte, _ jsonparser.ValueType, _ int) error {
+		if b, err := jsonparser.GetBoolean(v, "assetEnabled"); err == nil {
+			if b {
+				assetTypes = append(assetTypes, fmt.Sprintf("%q", asset))
+			}
+			e = jsonparser.Delete(e, "currencyPairs", "pairs", string(asset), "assetEnabled")
```

Comment:
I don't quite understand why this removes the field for every exchange. Why not keep the value?

Reply:
Version consistency.
If you downgrade, and it's left, then user would be left not knowing which was correct at the current version. They'd end up changing both, at best, or the wrong one at worst.
Do you disagree?

Reply:
If `null` becomes `false` and `true` remains `true` during upgrade, couldn't only `true` remain during downgrade?

Reply:
We discussed, and I think we're on the same page.
Partly, this is about idempotent upgrades and downgrades; If we upgrade twice, assetEnabled should remain there.
@gloriousCode 's concern was that upgrades takes the existing value of assetEnabled and translate null to false.
The reason for that is that this version is retrospective, and that we want defined and consistent behaviour on re-upgrading. So we need to deal with nulls and convert them to false.

So I think we're good.
TL;DR:
* Upgrade and downgrade must be Idempotent; No effect re-running
* Retrospective nature of this means it's possible it will encounter nulls, and it's good that it converts them
* Downgrade moving to a state with the old field removed does not create an unexpected situation

Reply:
My concern was in the other direction. When downgrading, this PR deletes existing fields rather than translating the value.
However, even if we were to only parse the `true` values when downgrading, GCT on running will still come in and add `true` fields to everything else on startup on v3. So I feel this instance is okay. I think `downgrade` should be able to transition between the versions just like `upgrade` and not just delete entire fields